mlir/lib/Support/TypeID.cpp

```diff
@@ -81,8 +81,11 @@ struct ImplicitTypeIDRegistry {
 } // end namespace
 
 TypeID detail::FallbackTypeIDResolver::registerImplicitTypeID(StringRef name) {
-  static ImplicitTypeIDRegistry registry;
-  return registry.lookupOrInsert(name);
+  // To prevent race conditions when one thread is accessing this `static`
+  // variable while other thread destructing it; construct the `registry`
+  // on the heap.
```

Comment:
```suggestion
  // To prevent race conditions when one thread is accessing this `static`
  // variable while the program is exiting and another thread already destroyed
  // it; intentionally construct the `registry` on the heap and never delete it.
```